sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/PaginationCarbonReader.java

```
@@ -0,0 +1,302 @@
+    // Initialize super class with no readers.
+    // Based on the splits identified for pagination query, readers will be built for the query.
+    super(null);
+    this.allBlockletSplits = splits;
+    this.readerBuilder = readerBuilder;
+    // prepare the mapping.
+    rowCountInSplits = new ArrayList<>(splits.size());
+    long sum = ((CarbonInputSplit) splits.get(0)).getDetailInfo().getRowCount();
+    rowCountInSplits.add(sum);
+    for (int i = 1; i < splits.size(); i++) {
+      // prepare a summation array of row counts in each blocklet,
+      // this is used for pruning with pagination vales.
+      // At current index, it contains sum of rows of all the blocklet from previous + current.
+      sum += ((CarbonInputSplit) splits.get(i)).getDetailInfo().getRowCount();
+      rowCountInSplits.add(sum);
+    }
+  }
+
+  /**
+   * Pagination query with from and to range.
+   *
+   * @param fromRowNumber must be greater 0 (as row id starts from 1)
+   *                      and less than equal to toRowNumber
+   * @param toRowNumber must be greater than 0 (as row id starts from 1)
+   *                and greater than equals to fromRowNumber and not outside the total rows
+   * @return array of rows between fromRowNumber and toRowNumber (inclusive)
+   * @throws Exception
+   */
+  public Object[] read(long fromRowNumber, long toRowNumber)
+      throws IOException, InterruptedException {
+    if (isClosed) {
+      throw new RuntimeException("Pagination Reader is closed. please build again");
+    }
+    if (fromRowNumber < 1) {
+      throw new IllegalArgumentException("from row id:" + fromRowNumber + " is less than 1");
+    }
+    if (fromRowNumber > toRowNumber) {
+      throw new IllegalArgumentException(
+          "from row id:" + fromRowNumber + " is greater than to row id:" + toRowNumber);
+    }
+    if (toRowNumber > getTotalRows()) {
+      throw new IllegalArgumentException(
```

here why do we need to throw an exception, if the limit is crossed, cant we give till the last row?

---

user should know what to pass, getTotalRows() provided for this reason. If by mistake he passes wrong value. we should not assume and give till end of rows.